core/templates/dev/head/components/HintAndSolutionButtonsDirective.js

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for hint and solution buttons.
+ */
+
+oppia.directive('hintAndSolutionButtons', [
+  'UrlInterpolationService', function(UrlInterpolationService) {
+    return {
+      restrict: 'E',
+      scope: {},
+      templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+        '/components/hint_and_solution_buttons_directive.html'),
+      controller: [
+        '$scope', '$rootScope', 'HintManagerService',
+        'ExplorationPlayerService', 'SolutionManagerService',
+        'PlayerTranscriptService', 'HintAndSolutionModalService',
+        'DeviceInfoService', 'WindowDimensionsService',
+        function(
+          $scope, $rootScope, HintManagerService,
+          ExplorationPlayerService, SolutionManagerService,
+          PlayerTranscriptService, HintAndSolutionModalService,
+          DeviceInfoService, WindowDimensionsService) {
+          $scope.defaultHintUrl = UrlInterpolationService.getStaticImageUrl(
+            '/icons/default_hint.svg');
+
+          $scope.activeHintUrl = UrlInterpolationService.getStaticImageUrl(
+            '/icons/active_hint.svg');
+
+          $scope.defaultSolutionUrl = UrlInterpolationService.getStaticImageUrl(
+            '/icons/default_solution.svg');
+
+          $scope.activeSolutionUrl = UrlInterpolationService.getStaticImageUrl(
+            '/icons/active_hint.svg');
+
+          $scope.hintPositions = [];
+          $scope.solutionExists = null;
+          $scope.isWindowNarrow = WindowDimensionsService.isWindowNarrow();
+          $scope.activeModalIndex = null;
+
+          $rootScope.$on('hintsAndSolutionReset', function(evt, data) {
+            for(var index = 0; index < data.numOfHints; index++) {
+              $scope.hintPositions.push(index);
+            }
+            $scope.solutionExists = data.solutionExists;
+          });
+
+          $scope.isHintButtonVisibleAtIndex = function(index) {
+            if (index === 0) {
+              return HintManagerService.isCurrentHintAvailable() ||
```

Comment:
The concept of a "current hint" doesn't really make sense any more; all "open" hints are current.

How about doing HintManagerService.getNumHintsReleased() and compare that with the given index? The HintManagerService can store a list of dicts (each dict has "numHintsReleased" and "solutionReleased" keys) that stores how many hints have been released for each card that's been shown so far. Note that card number (i.e. position of a card in the list of cards shown) is better than state name, since the player may loop back to a previous state.

Reply:
Concept of "current hint" still makes sense in the context of the desktop design. I think we can come up with a data structure once the design for both are finalized because handling both with the same data structure is a little confusing.

Reply:
Hm, I still think that "current hint" doesn't make sense in the mobile design (and I think you agree), so we shouldn't use it here. That said, I take your point about the desktop design needing the concept. There is also a need to cache "how many hints are open" status for previous cards which the current technical design doesn't really address.

So I would suggest updating to the newly-suggested data structure, and implementing getCurrentHint() etc. against that instead. (If I'm not mistaken, those functions correspond to surfacing a new hint or getting the latest hint for the current card, which seem like they would be straightforward to implement with the new design?) Let me know if I'm missing something.

Also, if anything, I see the desktop technical design moving towards this new representation as well, rather than the other way round. I think the new data structure suggested stores all the info that's needed.